File: python-algorithm/ds_5_tree/tree_1_2_inter_recurrent/L543.py
#--coding:utf-8--

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.res = -float("inf")
        self.transval(root)
        return self.res

    def transval(self, root):
        if not root: return 0
        lt_max = self.transval(root.left)
        rt_max = self.transval(root.right)
        self.res = max(self.res, lt_max+rt_max)
        return max(lt_max, rt_max)+1

File: python-algorithm/ds_5_tree/tree_1_2_inter_recurrent/L543.cpp


//二叉树的直径

#include <iostream>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int res = 0;
    int get(TreeNode* Node) {
        if (Node == NULL) return 0;
        //l,r代表的是每个当前节点的左右子树的高度
        int l = get(Node -> left);
        int r = get(Node -> right);
        //每个节点，都会做一次比较
        res = max(res, l+r);
        return max(l, r) + 1;
    }

    int diameterOfBinaryTree(TreeNode* root) {
        get(root);
        return res;
    }

};